### PR TITLE
Feature/hsl api endpoint

### PR DESCRIPTION
**Pull request by lanttu1243**
Add hsl api endpoint to website to allow serverside fetching to prevent API key from being out in the open and allow dynamic fetching so the ttimetable can be rendered client side.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,8 +22,13 @@
     "start": "dotenv -- turbo run start --cache-dir=.turbo",
     "typecheck": "turbo run typecheck --cache-dir=.turbo"
   },
+  "dependencies": {
```
Comment (MikaelSiidorow) [inline package.json:25]: These dependencies are probably accidentally added to root and should be added to the web and relevant packages instead

File: readme.md [hunk 2]
```diff
@@ -17,8 +17,8 @@ Prerequisites:
   NOTE: make sure you are installing MongoDB Database Tools, not something else, the documentation for Linux is quite wonky.
 
 ```sh
-# copy .env.example to .env
-cp .env.example .env
+# copy .env.secret.example to .env
+cp .env.secret.example .env
```
Comment (MikaelSiidorow) [inline readme.md:21]: Why was this changed, and the file deleted :eyes:

File: .idea/.gitignore
Comment (MikaelSiidorow) [inline .idea/.gitignore:1]: .idea folder should be added to .gitignore

File: apps/web/src/assets/HSL-API-types/hsl-helper-types.ts
Comment (MikaelSiidorow) [inline apps/web/src/assets/HSL-API-types/hsl-helper-types.ts:1]: assets sounds like something that would be used for e.g., images, videos and stuff, would rather move this to lib/types or another subdirectory

File: apps/web/src/app/HSL-API-endpoint/fetch-stops/route.ts [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+  ["HSL:2222603", "HSL:2222604"],
+  ["HSL:2222234", "HSL:2222212"],
+]; // [["HSL:2222234", "HSL:E22259"], ["HSL:2000102"], ["HSL:E0773", "HSL:E0772"]];
+
+export const dynamic = "force-dynamic";
+
+export async function GET() {
+  const dataFromHsl: RenderableStop[] = [];
+  for (const stop of STOPS) {
+    const stp: RenderableStop | null = await getStop(stop);
```
Comment (MikaelSiidorow) [inline apps/web/src/app/HSL-API-endpoint/fetch-stops/route.ts:26]: Do these need to be done sequentially? Or could we parallelize the requests for blazing fast speed 🔥

File: apps/web/src/app/HSL-API-endpoint/fetch-stops/route.ts [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    str = fullString.replace(subString, "");
+  }
+  return str;
+}
+
+function isTram(arrival: Arrival): boolean {
+  return arrival.route.includes("15");
+}
+function isMetro(arrival: Arrival): boolean {
+  return arrival.route.includes("M");
```
Comment (MikaelSiidorow) [inline apps/web/src/app/HSL-API-endpoint/fetch-stops/route.ts:86]: This seems a bit flaky? What if the route is Matinkylä or something else with a capital M

File: apps/web/src/app/HSL-API-endpoint/fetch-stops/route.ts [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+/*
+function checkIfPast(arrival: ArrivalAttribute): boolean {
+  const date = new Date();
+  const hour = date.getHours();
+  const min = date.getMinutes();
+  const currentTSM = hour * 60 + min;
+  const arrivalTime = Math.floor((arrival.hours * 60 + arrival.minutes) / 60);
+  return arrivalTime - currentTSM < -0.5;
+}
+*/
```
Comment (MikaelSiidorow) [inline apps/web/src/app/HSL-API-endpoint/fetch-stops/route.ts:170]:
```suggestion
```

Remove commented out code

**Review by MikaelSiidorow — CHANGES_REQUESTED** [review]
Some file cleanup, but otherwise looks good!

File: apps/web/src/app/HSL-API-endpoint/fetch-stops/route.ts (hunk 4 above)
Comment (lanttu1243) [reply apps/web/src/app/HSL-API-endpoint/fetch-stops/route.ts:86]: The arrival.route only includes the number of the bus i.e. 108N, 111, 52, 510 etc. so the M is only reserved for thee metro line.
arrival.headsign includes e.g. matinkylä and such. In the case the metro is replaced by a bus

File: apps/web/src/app/HSL-API-endpoint/fetch-stops/route.ts (hunk 3 above)
Comment (lanttu1243) [reply apps/web/src/app/HSL-API-endpoint/fetch-stops/route.ts:26]: It could be done but it fetches 10 arrivals for 6 stops in 42ms in localhost so I think the benefit of parallellism would be marginal but if it slows down significantly in production it can be implemented later.

Comment (kahlstrm) [reply apps/web/src/app/HSL-API-endpoint/fetch-stops/route.ts:26]: There is no reason here not to use [Promise.all](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all)/[Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) with a map inside instead of a for loop, don't know why ESLint is not complaining about the `await` inside the loop, I guess we have that rule disabled?

File: package.json [hunk 6]
```diff
@@ -24,6 +24,7 @@
   },
   "devDependencies": {
     "dotenv-cli": "^7.4.1",
+    "eslint": "^8.57.0",
```
Comment (kahlstrm) [inline package.json:27]: I don't think this is required here, as the subpackages contain `eslint` themselves and this ends up not being used. Or did you encounter some problem with the setup without this?

**Review by kahlstrm — CHANGES_REQUESTED** [review]
Even though now the files under `.idea` and the file `run_env.sh` are in `.gitignore`, they still need to be removed from git history.

I think moving the api endpoint under `next_api` would also make more sense, to not start a pattern of blowing up the middleware matcher regex too much.

**Review by kahlstrm — CHANGES_REQUESTED** [review]
image

I don't know what has happened here, but this PR is not ready to be reviewed, as the diff looks like almost the entire project has been added.
Please rebase this on top of main somehow so the diff looks sensible, probably it is easiest to create a new branch from `main` and then cherry-pick your commits one by one on it, and then git reset hard this branch to that one.

File: turbo.json [hunk 7]
```diff
@@ -3,6 +3,7 @@
   "globalDependencies": ["**/.env.*local"],
   "globalDotEnv": [".env"],
   "globalEnv": [
+    "DIGITRANSIT_SUBSCRIPTION_KEY",
```
Comment (kahlstrm) [inline turbo.json:6]: Add this to `.env.example` with a comment of what it is for?

File: apps/web/src/app/next_api/fetch-hsl-stops/route.ts [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+  second: string,
+  n = N_ARRIVALS,
+): Promise<RenderableStop | null> => {
+  let data: Stop | null = null;
+  let data2: Stop | null = null;
+  await getData(first).then((result: Stop | null) => (data = result));
+  await getData(second).then((result: Stop | null) => (data2 = result));
+
+  const result1: StopOutData | null = toOutData(data);
+  const result2: StopOutData | null = toOutData(data2);
```
Comment (kahlstrm) [inline apps/web/src/app/next_api/fetch-hsl-stops/route.ts:208]:
```suggestion
const getStop = async (
  stops: readonly [string, string],
  n = N_ARRIVALS,
): Promise<RenderableStop | null> => {
  const [result1, result2] = await Promise.all(
    stops.map((stop) => getData(stop).then(toOutData)),
  );
```

**Review by kahlstrm — CHANGES_REQUESTED** [review]

**Review by kahlstrm — APPROVED** [review]
Looks good, left one comment

**Comment by kahlstrm** [comment]
As this is not being used anywhere (yet), I think we probably don't want to merge this before we have the UI-functionality added as well. Can you do a separate PR for that that is built on top of this one?

File: apps/web/src/app/next_api/fetch-hsl-stops/route.ts [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+    " via Tapiola",
+    " via Pasila as.",
+  ];
+  let str = fullString;
+  for (const subString of subStrings) {
+    // HSL sometimes has a bug where HEadSign is null so this handles case string in is null :D
+    if (str) {
+      str = str.replace(subString, "");
+    } else {
+      ("Null");
```
Comment (MikaelSiidorow) [inline apps/web/src/app/next_api/fetch-hsl-stops/route.ts:97]: Is this line doing anything? It doesn't look like the string value is returned or used at all, could this else branch be removed completely?

File: apps/web/src/app/next_api/fetch-hsl-stops/route.ts [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+      .sort((arr1, arr2) => arr1.realTimeArrival - arr2.realTimeArrival)
+      .slice(0, n + 1),
+  };
+  const ArrivalsFormatted: ArrivalAttribute[] = result.arrival
+    .map((arr: Arrival) => {
+      return {
+        route: arr.route ? arr.route.replace(" ", "") : "Null",
+        headSign: removeSubstring(arr.headSign),
+        hours:
+          Math.floor((arr.realTimeArrival - arr.serviceDay) / 60 / 60) % 24,
```
Comment (MikaelSiidorow) [inline apps/web/src/app/next_api/fetch-hsl-stops/route.ts:217]: If the API response times are in any reasonable time format (timestamp or ISO etc.), we could probably use more native datetime operations here, but this mankelointi is probably fine for now.

**Review by MikaelSiidorow — APPROVED** [review]
🚀  Looks good

Having external APIs with GraphQL only endpoints should be illegal, but there's nothing we can do about that

**Review by kahlstrm — COMMENTED** [review]
Hmm, I think this commit is doing the wrong stuff no? 😅
